### ISSUE TITLE
How to run val.py on My Custom Test Dataset?

### ISSUE THREAD

**Issue by joneswilliam1** [ultralytics/yolov5#7852]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi everyone,

May I know, how do I run validation on my test dataset? I thought that I could do so by using the ```--task``` argument, however, when I provide the following command:
```
python val.py --weights best.pt --data data.yaml --task test
```

My data.yaml is as follows:
```
train: train/images
val: valid/images/tes
test: test/images

nc: 4
names: ['bus', 'car', 'motorcycle', 'truck']
```

It ends up searching however val path and gives the Exception "Dataset not found". Why does it search in the "Val" Path when I provide the argument ```task``` with "test"?

Thanks.

**Comment by glenn-jocher** [comment]
@joneswilliam1 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. Your issue is not reproducible. val.py --task test will use your data yaml `test:` key.


We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by abuelgasimsaadeldin** [comment]
@glenn-jocher Here are the steps to reproduce:
1) Clone and Git Pull latest YOLOv5 Repo
2) Place Folder "test" in root directory with folders "images" and "labels" inside
3) Add data.yaml file in root directory as shown below:
```
train: train/images
val: valid/images
test: test/images

nc: 4
names: ['bus', 'car', 'motorcycle', 'truck']
```
![image](https://user-images.githubusercontent.com/68045710/168782432-73539092-2580-4ef0-9a91-2519a845b752.png)


4) Run Following Command: ```python val.py --weights best.pt --data data.yaml --task test```

5) Result is shown in Snapshot below:
![img](https://user-images.githubusercontent.com/68045710/168783129-bf535f6b-01cc-4774-bc22-6332f8dfc50f.png)

**Comment by joneswilliam1** [comment]
Yes @abuelgasimsaadeldin, that is exactly the error which I am getting. Thanks for that!

**Comment by glenn-jocher** [comment]
@joneswilliam1 @abuelgasimsaadeldin 👋 Hello! Thanks for asking about YOLOv5 🚀 **dataset formatting**. To train correctly your data must be in YOLOv5 format. Please see our [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) tutorial for full documentation on dataset setup and all steps required to start training your first model. A few excerpts from the tutorial:

### 1.1 Create dataset.yaml

[COCO128](https://www.kaggle.com/datasets/ultralytics/coco128) is an example small tutorial dataset composed of the first 128 images in [COCO](http://cocodataset.org/#home) train2017. These same 128 images are used for both training and validation to verify our training pipeline is capable of overfitting. [data/coco128.yaml](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), shown below, is the dataset config file that defines 1) the dataset root directory `path` and relative paths to `train` / `val` / `test` image directories (or *.txt files with image paths), 2) the number of classes `nc` and 3) a list of class `names`:
```yaml
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: ../datasets/coco128  # dataset root dir
train: images/train2017  # train images (relative to 'path') 128 images
val: images/train2017  # val images (relative to 'path') 128 images
test:  # test images (optional)

# Classes
nc: 80  # number of classes
names: [ 'person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
         'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
         'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
         'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
         'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
         'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
         'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
         'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
         'hair drier', 'toothbrush' ]  # class names
```


### 1.2 Create Labels

After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:

- One row per object
- Each row is `class x_center y_center width height` format.
- Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
- Class numbers are zero-indexed (start from 0).

<img width="800" alt="Image Labels" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg">

The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):

<p align="center"><img width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png"></p>


### 1.3 Organize Directories

Organize your train and val images and labels according to the example below. YOLOv5 assumes  `/coco128` is inside a `/datasets` directory **next to** the `/yolov5` directory. **YOLOv5 locates labels automatically for each image** by replacing the last instance of `/images/` in each image path with `/labels/`. For example:
```bash
../datasets/coco128/images/im0.jpg  # image
../datasets/coco128/labels/im0.txt  # label
```

<p align="center"><img width="700" src="https://user-images.githubusercontent.com/26833433/134436012-65111ad1-9541-4853-81a6-f19a3468b75f.png"></p>
</details>


Good luck 🍀 and let us know if you have any other questions!

**Comment by abuelgasimsaadeldin** [comment]
@glenn-jocher I see, seems like I was just missing the ```path: .``` at the start. Sorry about that and thanks! @joneswilliam1 it seems to be working just fine now..

**Comment by EslamHaourn** [comment]
Hi everyone,
I have a question please,
If I have small dataset and split it only into train set and val set as following:
train: ../kitti(augment3classes_crop)/images/train/  # 5751 images
val: ../kitti(test_augment3classes_crop)/images/train/  # 1730 images
test: ../kitti(test_augment3classes_crop)/images/train/ # 1730 images
To evaluate my model and get mAP for val set I run val.py with test task. The val.py file doesn't work without the third line start with test. 
In training, the mAP values appears for 1730 images which is the val set, Should the dataset used for evaluating model havn't seen before?
Is this correct? If not what should I do?
Thanks.

**Comment by glenn-jocher** [comment]
@EslamHaourn validating on val set is fine

**Comment by EslamHaourn** [comment]
> @EslamHaourn validating on val set is fine

Thanks for your reply.

How could I evaluate my model after the second and the last portion of data (val set) used in training?

I got an overfit with dropout when I set my data as following:
train: ../kitti(augment3classes_crop)/images/train/ # 5751 images
val: ../kitti(augment3classes_crop)/images/train/ # 5751 images
test: ../kitti(test_augment3classes_crop)/images/train/ # 1730 images
where the portion used for evaluting the model (test set) not used in training.

**Comment by glenn-jocher** [comment]
```
python val.py --task test
```
